app/lib/actions/katello/environment/destroy.rb

```diff
@@ -24,14 +24,21 @@ def plan(env, options = {})
             end
 
             if organization_destroy
-              env.hostgroups.clear
-              env.hosts.clear
+              delete_hosts_and_hostgroups(environment: env)
             end
 
             plan_self
           end
         end
 
+        def delete_hosts_and_hostgroups(environment:)
+          environment.hostgroups.delete_all
+          host_ids = environment.hosts.ids
+          ::Katello::Host::ContentFacet.where(:host_id => host_ids).delete_all
+          ::Katello::Host::SubscriptionFacet.where(:host_id => host_ids).delete_all
+          ::Host::Managed.where(:id => host_ids).delete_all
```

Comment:
Don't delete the hosts. The default behavior today is if you delete the org the host remains unassigned but still visible via 'Any Organization' => 'All Hosts'

```suggestion
```

Reply:
The previous code was 

```rb
env.hosts.clear
```

[This](https://apidock.com/rails/ActiveRecord/Associations/CollectionProxy/clear) says `clear` is equivalent to `delete_all`. Wouldn't that also delete the hosts?

Reply:
They may be equivalent but remember that env to host association looked lie this
```ruby
 has_many :hosts,      :class_name => "::Host::Managed", :through => :content_facets,
                          :inverse_of => :lifecycle_environment
```

The `env.hosts.clear` would delete the associated content facets not the host itself.

Reply:
`env.hosts.clear` is equivalent to `env.hosts.delete_all` which would indeed delete the hosts. I'm not understanding what you're saying here..

Reply:
TIL that `delete_all` and `clear` don't delete the named records if they're called at the end of a chain. wow.

Updated to not call `delete_all` for hosts (also fixed it in the other location)